File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int ans = 0;
        int n = height.size();

        vector<int>leftMax;
        vector<int>rightMax(n);
        
        int curr = INT_MIN;
        for(int &i:height){
           curr = max(curr,i);
           leftMax.push_back(curr);
        }
        curr=INT_MIN;
        for(int i=n-1;i>=0;i--){
             curr = max(height[i],curr);
             rightMax[i] = curr;
        }

        for(int i=0;i<n;i++){
            ans+=min(rightMax[i],leftMax[i])-height[i];
        }
        return ans;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.java

class Solution {
    private int dynamicProgramming(int[] height){
        int n = height.length;
        int ans = 0;

        int[] leftMax = new int[n];
        int[] rightMax = new int[n];
        
        int lMax = -1;
        int rMax = -1;
        for(int i=0;i<n;i++){
          lMax = Math.max(lMax,height[i]);
          rMax = Math.max(rMax,height[n-1-i]);

          leftMax[i]=lMax;
          rightMax[n-1-i]=rMax;
        }

        for(int i=0;i<n;i++){
            ans+= Math.min(leftMax[i],rightMax[i]) - height[i];
        }

        return ans;
    }

    int twoPointer(int[] height){
          int left = 0, right = height.length - 1;
          int ans = 0, leftMax = 0, rightMax = 0;

          while(left < right){
             if(height[left]<height[right]){
                leftMax = Math.max(leftMax,height[left]);
                ans+= leftMax - height[left];
                ++left;
             }
             else {
                rightMax = Math.max(rightMax,height[right]);
                ans+= rightMax - height[right];
                --right;
             }
          }

          return ans; 
    }
    public int trap(int[] height) {
        return twoPointer(height);
    }
}